### PR TITLE
Created a discovery handler using asyncio

### PR DESCRIPTION
**Pull request by rokam**
I did create a MideaDiscovery class that handles discovery with some async events. With this, we'll be able to automatically support discovery inside homeassistant.

I wasn't able to test it with v1 and v3 devices, I'm glad if someone test it for me.

### CODE REVIEW

**Comment by seppe912** [comment]
got error `Error : (<class 'TypeError'>, TypeError('support_testv3() takes 3 positional arguments but 4 were given'), <traceback object at 0x75eefb48>)`

need to remove `self` on scanner.py line 39:
_device = await self.support_testv3(account, password)
then OK with my V3

**Comment by rokam** [comment]
> got error `Error : (<class 'TypeError'>, TypeError('support_testv3() takes 3 positional arguments but 4 were given'), <traceback object at 0x75eefb48>)`
> 
> need to remove `self` on scanner.py line 39:
> _device = await self.support_testv3(account, password)
> then OK with my V3

Thanks. Fixed

File: msmart/cli.py [hunk 1]
```diff
@@ -25,230 +16,35 @@
...
-    b1, b2 = b[:16], b[16:]
-    b3 = bytearray(16)
-    i = 0
-    while i < len(b1):
-        b3[i] = b1[i] ^ b2[i]
-        i += 1
-    return b3
+    try:
+        discovery = MideaDiscovery(account=account, password=password)
+        found_devices = asyncio.run(discovery.get_all())
```
Comment (0xbw) [inline msmart/cli.py]: `asyncio.run` is a Python 3.7 addition. 
You need to check the version of python in line 10

File: msmart/cli.py [hunk 2]
```diff
@@ -1,20 +1,11 @@
...
-from msmart.security import get_udpid
-from msmart.device import air_conditioning_device as ac
-from msmart.cloud import cloud
-try:
-    import xml.etree.cElementTree as ET
-except ImportError:
-    import xml.etree.ElementTree as ET
+from msmart.scanner import MideaDiscovery
 
 if sys.version_info < (3, 5):
```
Comment (0xbw) [inline msmart/cli.py:10]:
```python
if sys.version_info < (3, 7):
```

File: msmart/cli.py (hunk 1 above)
Comment (rokam) [reply msmart/cli.py]: I've changed to support python 3.5

**Comment by rokam** [comment]
@seppe912 I did some changes, can you please test it again with v3 devices?

File: msmart/scanner.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+async def _get_networks():
+    nets :list() = []
+    adapters = ifaddr.get_adapters()
+    for adapter in adapters:
+        for ip in adapter.ips:
+            localNet = None
+            if ip.is_IPv4:
+                localNet = ipaddress.IPv4Network(f"{ip.ip}/{ip.network_prefix}", strict=False)
+            elif ip.is_IPv6:
```
Comment (0xbw) [inline msmart/scanner.py:249]: no need to check ipv6, ipv6 does not support broadcast addresses, and midea  does not support ipv6

File: msmart/scanner.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    def __str__(self):
+        return str(self.__dict__)
+    
+    async def support_test(self, account=OPEN_MIDEA_APP_ACCOUNT, password=OPEN_MIDEA_APP_PASSWORD):
+        if self.version == 3:
+            _device = await self.support_testv3(account, password)
+        else:
+            _device = ac(self.ip, self.id, self.port)
+        if self.type == 'ac':
+            loop = asyncio.get_running_loop()
```
Comment (0xbw) [inline msmart/scanner.py:46]: you may use `asyncio.get_event_loop()`

File: msmart/scanner.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    async def _get_response(self, ip=None):
+        try:
+            data, addr = self.socket.recvfrom(512)
+            if ip is not None and addr[0] != ip:
+                return None
+            else:
+                ip = addr[0]
+            _LOGGER.debug("Midea Local Data {} {}".format(ip, data.hex()))
+            device = await scandevice.load(ip, data)
+            return asyncio.create_task(device.support_test(self.account, self.password))
```
Comment (0xbw) [inline msmart/scanner.py:198]:
```python
                loop = asyncio.get_event_loop()
                return loop.create_task(device.support_test(self.account, self.password))
```

File: msmart/scanner.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    nets :list() = []
+    adapters = ifaddr.get_adapters()
+    for adapter in adapters:
+        for ip in adapter.ips:
+            localNet = None
+            if ip.is_IPv4:
+                localNet = ipaddress.IPv4Network(f"{ip.ip}/{ip.network_prefix}", strict=False)
+            elif ip.is_IPv6:
+                localNet = ipaddress.IPv6Network(f"{ip.ip[0]}/{ip.network_prefix}", strict=False)
+            if localNet.is_private and not localNet.is_loopback and not localNet.is_link_local and len(localNet.hosts) > 1:
```
Comment (0xbw) [inline msmart/scanner.py:251]: `TypeError: object of type 'method' has no len()`

File: msmart/scanner.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+            await loop.run_in_executor(None, _device.refresh)
+            _LOGGER.debug("{}".format(_device))
+            self.support = _device.support
+        _LOGGER.info("*** Found a device: \033[94m\033[1m{} \033[0m".format(self)) 
+        return self
+
+    async def support_testv3(self, account, password):
+        _device = ac(self.ip, self.id, self.port)
+        for udpid in [get_udpid(self.id.to_bytes(6, 'little')), get_udpid(self.id.to_bytes(6, 'big'))]:
+            loop = asyncio.get_running_loop()
```
Comment (0xbw) [inline msmart/scanner.py:56]:
```python
            loop = asyncio.get_event_loop()
            token, key = await loop.run_in_executor(None, gettoken, udpid, account, password)
            auth = await loop.run_in_executor(None, _device.authenticate, key, token)
```

**Comment by 0xbw** [comment]
@rokam Sorry, please reopen a PR. 😄

**Comment by rokam** [comment]
Ok

**Pull request merged**